File: Components/WinForms/VB/EditorsMainDemo/Modules/ModuleListBox.vb
Imports System
Imports System.Collections
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports DevExpress.Tutorials
Imports System.Windows.Forms
' <imageListBoxControlSample>
Imports DevExpress.XtraEditors
' </imageListBoxControlSample>

Namespace DevExpress.XtraEditors.Demos
	Partial Public Class ModuleListBox
		Inherits TutorialControl

		Public Sub New()
			InitializeComponent()
            InitListBox()
            InitValues()
            TutorialInfo.WhatsThisCodeFile = "VB\EditorsMainDemo\Modules\ModuleListBox.vb;VB\EditorsMainDemo\Modules\ListBoxProperties.vb"
			TutorialInfo.WhatsThisXMLFile = "listbox.xml"
        End Sub
        Private scaleFactor As SizeF = New SizeF(1.0F, 1.0F)
        Protected Overrides Sub ScaleControl(factor As SizeF, specified As BoundsSpecified)
            MyBase.ScaleControl(factor, specified)
            Me.scaleFactor = factor
        End Sub

		Protected Overrides ReadOnly Property AllowCenterControls() As Boolean
			Get
				Return False
			End Get
		End Property
		Private properties As ListBoxProperties = Nothing

        Private ReadOnly Property CurrentListBox() As BaseListBoxControl
            Get
                Select Case xtraTabControl1.SelectedTabPageIndex
                    Case 1
                        Return checkedListBoxControlSample
                    Case 2
                        Return imageListBoxControlSample
                End Select
                Return listBoxControlSample
            End Get
        End Property


		Private Sub InitValues()
			If properties Is Nothing Then
				properties = New ListBoxProperties(CurrentListBox)
				TurnSearch(True)
				AddHandler properties.SearchControlPropertyChanged, AddressOf properties_SearchControlPropertyChanged
				properties.Dock = DockStyle.Fill
				xtraTabPage5.Controls.Add(properties)
			Else
				properties.InitListBox(CurrentListBox)
			End If
		End Sub

		Private Sub properties_SearchControlPropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
			Dim args As PropertyValueChangedEventArgs = TryCast(e, PropertyValueChangedEventArgs)
			Select Case args.PropertyName
				Case "AllowAutoApply"
					' <ceAllowAutoApply>
					searchControlSample.Properties.AllowAutoApply = CBool(args.Value)
					' </ceAllowAutoApply>
					Return
				Case "ShowSearchButton"
					' <ceShowSearchButton>
					searchControlSample.Properties.ShowSearchButton = CBool(args.Value)
					' </ceShowSearchButton>
					Return
				Case "ShowClearButton"
					' <ceShowClearButton>
					searchControlSample.Properties.ShowClearButton = CBool(args.Value)
					' </ceShowClearButton>
					Return
                Case "ShowMRUButton"
                    ' <ceShowMRUButton>
                    searchControlSample.Properties.ShowMRUButton = CBool(args.Value)
                    ' </ceShowMRUButton>
                    Return
                Case "FindDelay"
                    ' <tbFindDelay>
                    searchControlSample.Properties.FindDelay = CInt(Fix(args.Value))
                    ' </tbFindDelay>
                    Return
                Case "FilterCondition"
                    ' <cbFilterCondition>
                    searchControlSample.Properties.FilterCondition = CType(args.Value, DevExpress.Data.Filtering.FilterCondition)
                    ' </cbFilterCondition>
                    Return
                Case "ShowDefaultButtonsMode"
                    ' <cbShowDefaultButtonsMode>
                    searchControlSample.Properties.ShowDefaultButtonsMode = CType(args.Value, Repository.ShowDefaultButtonsMode)
                    ' <cbShowDefaultButtonsMode>
                Case "TurnSearch"
                    splitContainerControl.PanelVisibility = If(CBool(args.Value), SplitPanelVisibility.Both, SplitPanelVisibility.Panel2)
                    ' <searchControlSample>
                    TurnSearch(CBool(args.Value))
                    ' </searchControlSample>
                    Return
            End Select
		End Sub
		' <searchControlSample>
		Private Sub TurnSearch(ByVal allowSearch As Boolean)
			Dim listBoxControl As ListBoxControl = TryCast(CurrentListBox, ListBoxControl)
			If allowSearch AndAlso listBoxControl IsNot Nothing Then
				searchControlSample.Client = listBoxControl
			Else
				searchControlSample.Client = Nothing
			End If
		End Sub
		' </searchControlSample>        

		Private Sub InitListBox()
			listBoxControlSample.DataSource = TutorialHelper.Countries
			'
			' <checkedListBoxControlSample>
			checkedListBoxControlSample.DataSource = GetData()
			If checkedListBoxControlSample.DataSource IsNot Nothing Then
				checkedListBoxControlSample.DisplayMember = "ProductName"
				checkedListBoxControlSample.ValueMember = "Discontinued"
				Dim i As Integer = 0
				Do While checkedListBoxControlSample.GetItem(i) IsNot Nothing
					checkedListBoxControlSample.SetItemCheckState(i, (If(True.Equals(checkedListBoxControlSample.GetItemValue(i)), CheckState.Unchecked, CheckState.Checked)))
					i += 1
				Loop
			End If
			' </checkedListBoxControlSample>
            '
            Dim imageSize As Size = New Size(Convert.ToInt32(25 * scaleFactor.Width), Convert.ToInt32(16 * scaleFactor.Height))
            DevExpress.Tutorials.TutorialHelper.InitFont(imageListBoxControlSample, imageSize)
			imageListBoxControlSample.SelectedIndex = 0
		End Sub

		' <checkedListBoxControlSample>
		Private Function GetData() As DataTable
			Dim DBFileName As String = DevExpress.Utils.FilesHelper.FindingFileName(Application.StartupPath, "Data\nwind.mdb")
			If DBFileName <> "" Then
				Dim ds As New DataSet()
				Dim con As String = "Provider=Microsoft.Jet.OLEDB.4.0;Data Source=" & DBFileName

				Dim dbAdapter As New System.Data.OleDb.OleDbDataAdapter("SELECT * FROM Products", con)
				dbAdapter.Fill(ds, "Products")
				Return ds.Tables("Products")
			End If
			Return Nothing
		End Function
		' </checkedListBoxControlSample>


		Private Sub xtraTabControl1_SelectedPageChanged(ByVal sender As Object, ByVal e As DevExpress.XtraTab.TabPageChangedEventArgs) Handles xtraTabControl1.SelectedPageChanged
			InitValues()
		End Sub

		' <ceCustomDraw>
		Private Sub checkedListBoxControlSample_DrawItem(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.ListBoxDrawItemEventArgs) Handles checkedListBoxControlSample.DrawItem
			If Not ceCustomDraw.Checked Then
				Return
			End If
			If checkedListBoxControlSample.GetItemChecked(e.Index) Then
				Return
			End If
			e.Appearance.Font = New Font(e.Appearance.Font, FontStyle.Strikeout)
		End Sub

		Private Sub ceCustomDraw_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles ceCustomDraw.CheckedChanged
			checkedListBoxControlSample.Refresh()
		End Sub
		' </ceCustomDraw>

		' <imageListBoxControlSample>
		Private Sub imageListBoxControlSample_SelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles imageListBoxControlSample.SelectedIndexChanged
			Dim ilb As ImageListBoxControl = TryCast(sender, ImageListBoxControl)
			Dim fontName As Object = ilb.SelectedValue
			lbSample.Visible = fontName IsNot Nothing
			If fontName Is Nothing Then
				Return
			End If
			lbSample.Font = New Font(fontName.ToString(), 12)
		End Sub
		' </imageListBoxControlSample>

		' <seInvert>
		Private Sub seInvert_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles seInvert.Click
			checkedListBoxControlSample.BeginUpdate()
			Try
				Dim i As Integer = 0
				Do While checkedListBoxControlSample.GetItem(i) IsNot Nothing
					checkedListBoxControlSample.SetItemCheckState(i, (If(checkedListBoxControlSample.GetItemChecked(i), CheckState.Unchecked, CheckState.Checked)))
					i += 1
				Loop
			Finally
				checkedListBoxControlSample.EndUpdate()
			End Try
		End Sub
		' </seInvert>
	End Class
End Namespace

File: Components/WinForms/CS/EditorsMainDemo/Modules/ModuleListBox.cs
using System;
using System.Collections;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using DevExpress.Tutorials;
using System.Windows.Forms;
// <imageListBoxControlSample>
using DevExpress.XtraEditors;
// </imageListBoxControlSample>

namespace DevExpress.XtraEditors.Demos {
    public partial class ModuleListBox : TutorialControl {
        public ModuleListBox() {
            InitializeComponent();
            InitListBox();
            InitValues();
            TutorialInfo.WhatsThisCodeFile = "CS\\EditorsMainDemo\\Modules\\ModuleListBox.cs;CS\\EditorsMainDemo\\Modules\\ListBoxProperties.cs";
            TutorialInfo.WhatsThisXMLFile = "DevExpress.XtraEditors.Demos.CodeInfo.listbox.xml";
        }
        SizeF scaleFactor = new SizeF(1f, 1f);
        protected override void ScaleControl(SizeF factor, BoundsSpecified specified) {
            base.ScaleControl(factor, specified);
            this.scaleFactor = factor;
        }
        protected override bool AllowCenterControls { get { return false; } }
        ListBoxProperties properties = null;

        BaseListBoxControl CurrentListBox {
            get {
                switch(xtraTabControl1.SelectedTabPageIndex) {
                    case 1: return checkedListBoxControlSample;
                    case 2: return imageListBoxControlSample;
                }
                return listBoxControlSample;
            }
        }
                
        void InitValues() {
            if(properties == null) {
                properties = new ListBoxProperties(CurrentListBox);
                TurnSearch(true);
                properties.SearchControlPropertyChanged += new PropertyChangedEventHandler(properties_SearchControlPropertyChanged);                
                properties.Dock = DockStyle.Fill;
                xtraTabPage5.Controls.Add(properties);
            }
            else {
                properties.InitListBox(CurrentListBox);
            }
        }

        void properties_SearchControlPropertyChanged(object sender, PropertyChangedEventArgs e) {
            PropertyValueChangedEventArgs args = e as PropertyValueChangedEventArgs;
            switch(args.PropertyName) {
                case "AllowAutoApply":
                    // <ceAllowAutoApply>
                    searchControlSample.Properties.AllowAutoApply = (bool)args.Value;
                    // </ceAllowAutoApply>
                    return;
                case "ShowSearchButton":
                    // <ceShowSearchButton>
                    searchControlSample.Properties.ShowSearchButton = (bool)args.Value;
                    // </ceShowSearchButton>
                    return;
                case "ShowClearButton":
                    // <ceShowClearButton>
                    searchControlSample.Properties.ShowClearButton = (bool)args.Value;
                    // </ceShowClearButton>
                    return;
                case "ShowMRUButton":
                    // <ceShowMRUButton>
                    searchControlSample.Properties.ShowMRUButton = (bool)args.Value;
                    // </ceShowMRUButton>
                    return;
                case "FindDelay":
                    // <tbFindDelay>
                    searchControlSample.Properties.FindDelay = (int)args.Value;
                    // </tbFindDelay>
                    return;
                case "FilterCondition":
                    // <cbFilterCondition>
                    searchControlSample.Properties.FilterCondition = (DevExpress.Data.Filtering.FilterCondition)args.Value;
                    // </cbFilterCondition>
                    return;
                case "ShowDefaultButtonsMode":
                    // <cbShowDefaultButtonsMode>
                    searchControlSample.Properties.ShowDefaultButtonsMode = (Repository.ShowDefaultButtonsMode)args.Value;
                    // <cbShowDefaultButtonsMode>
                    break;
                case "TurnSearch":
                    splitContainerControl.PanelVisibility = (bool)args.Value ? SplitPanelVisibility.Both : SplitPanelVisibility.Panel2;
                    // <searchControlSample>
                    TurnSearch((bool)args.Value);
                    // </searchControlSample>
                    return;
            }
        }        
        // <searchControlSample>
        void TurnSearch(bool allowSearch) {            
            ListBoxControl listBoxControl = CurrentListBox as ListBoxControl;
            if(allowSearch && listBoxControl != null)
                searchControlSample.Client = listBoxControl;
            else
                searchControlSample.Client = null;            
        }
        // </searchControlSample>        

        void InitListBox() {
            listBoxControlSample.DataSource = TutorialHelper.Countries;
            //
            // <checkedListBoxControlSample>
            checkedListBoxControlSample.DataSource = GetData();
            if(checkedListBoxControlSample.DataSource != null) {
                checkedListBoxControlSample.DisplayMember = "ProductName";
                checkedListBoxControlSample.ValueMember = "Discontinued";
                int i = 0;
                while(checkedListBoxControlSample.GetItem(i) != null) {
                    checkedListBoxControlSample.SetItemCheckState(i, (true.Equals(checkedListBoxControlSample.GetItemValue(i)) ? CheckState.Unchecked : CheckState.Checked));
                    i++;
                }
            }
            // </checkedListBoxControlSample>
            //
            Size imageSize = new System.Drawing.Size((int)(25 * scaleFactor.Width), (int)(16 * scaleFactor.Height));
            DevExpress.Tutorials.TutorialHelper.InitFont(imageListBoxControlSample, imageSize);
            imageListBoxControlSample.SelectedIndex = 0;
        }

        // <checkedListBoxControlSample>
        DataTable GetData() {
            string DBFileName = DevExpress.Utils.FilesHelper.FindingFileName(Application.StartupPath, "Data\\nwind.mdb");
            if(DBFileName != "") {
                DataSet ds = new DataSet();
                string con = "Provider=Microsoft.Jet.OLEDB.4.0;Data Source=" + DBFileName;

                System.Data.OleDb.OleDbDataAdapter dbAdapter = new System.Data.OleDb.OleDbDataAdapter("SELECT * FROM Products", con);
                dbAdapter.Fill(ds, "Products");
                return ds.Tables["Products"];
            }
            return null;
        }
        // </checkedListBoxControlSample>


        private void xtraTabControl1_SelectedPageChanged(object sender, DevExpress.XtraTab.TabPageChangedEventArgs e) {
            InitValues();
        }

        // <ceCustomDraw>
        private void checkedListBoxControlSample_DrawItem(object sender, DevExpress.XtraEditors.ListBoxDrawItemEventArgs e) {
            if(!ceCustomDraw.Checked) return;
            if(checkedListBoxControlSample.GetItemChecked(e.Index)) return;
            e.Appearance.Font = new Font(e.Appearance.Font, FontStyle.Strikeout);
        }

        private void ceCustomDraw_CheckedChanged(object sender, System.EventArgs e) {
            checkedListBoxControlSample.Refresh();
        }
        // </ceCustomDraw>

        // <imageListBoxControlSample>
        private void imageListBoxControlSample_SelectedIndexChanged(object sender, System.EventArgs e) {
            ImageListBoxControl ilb = sender as ImageListBoxControl;
            object fontName = ilb.SelectedValue;
            lbSample.Visible = fontName != null;
            if(fontName == null) return;
            lbSample.Font = new Font(fontName.ToString(), 12);
        }
        // </imageListBoxControlSample>

        // <seInvert>
        private void seInvert_Click(object sender, System.EventArgs e) {
            checkedListBoxControlSample.BeginUpdate();
            try {
                int i = 0;
                while(checkedListBoxControlSample.GetItem(i) != null) {
                    checkedListBoxControlSample.SetItemCheckState(i, (checkedListBoxControlSample.GetItemChecked(i) ? CheckState.Unchecked : CheckState.Checked));
                    i++;
                }
            }
            finally {
                checkedListBoxControlSample.EndUpdate();
            }
        }
        // </seInvert>
    }
}

